### PR TITLE
add 'map::clear()', 'map::has()', 'map::insert()', 'map::set()' convenience methods

### PR DESCRIPTION
**Pull request by jwerle**

### CODE REVIEW

File: modules/map/src/lib.zz [hunk 1]
```diff
@@ -407,6 +407,46 @@ export fn next(Keys mut* self) -> bool
     return false;
 }
 
+/// Clear all references for all keys in the `Map`. Returns `true` upon
+/// success, otherwise `false`.
+export fn clear(Map mut *self) bool {
```
Comment (aep) [inline modules/map/src/lib.zz:412]: this can't fail. no need to return bool

File: modules/map/src/lib.zz [hunk 2]
```diff
@@ -407,6 +407,46 @@ export fn next(Keys mut* self) -> bool
...
+
+  return true;
+}
+
+/// Returns true if there is a value entry for a given key, otherwise
+/// false.
+export fn has(Map *self, void *key) bool
+{
+  let value = self->get(key);
+  if 0 != self->get(key) {
```
Comment (aep) [inline modules/map/src/lib.zz:427]: duplicated call to get()

**Review by aep — CHANGES_REQUESTED** [review]

File: modules/map/src/lib.zz [hunk 3]
```diff
@@ -407,6 +407,42 @@ export fn next(Keys mut* self) -> bool
     return false;
 }
 
+/// Clear all references for all keys in the `Map`.
+export fn clear(Map mut *self)
+{
+  for let mut it = self->keys(); it.next(); {
+    self->remove(it.key.mem);
```
Comment (aep) [inline modules/map/src/lib.zz:414]: actually, this should probably avoid rehashing the keys.

look at remove_slk for iterating the internal structure

**Comment by aep** [comment]
@jwerle  in line with #132 i believe insert should copy while set does borrow
i.e. the naming convention is "the more forceful verb does a copy"

**Comment by jwerle** [comment]
Makes sense!

**Comment by jwerle** [comment]
Still working on this! Should have out this morning

File: modules/map/src/main.zz [hunk 4]
```diff
@@ -38,11 +38,18 @@ pub fn main() -> int {
...
+
+    err::assert(amap.count() == 3);
 
     let gimmeback = (A*)amap.get("bob");
     err::assert_safe(gimmeback);
     static_attest(buffer::integrity(&gimmeback->b));
     log::info("%s", gimmeback->b.cstr());
+    log::info("%lu", amap.count());
+    amap.clear();
+    err::assert(amap.count() == 0);
```
Comment (aep) [inline modules/map/src/main.zz:52]: could you add a test please that checks that the pool memory has been returned?

i.e. compare pool::free_bytes()

File: modules/map/src/lib.zz [hunk 5]
```diff
@@ -407,6 +407,61 @@ export fn next(Keys mut* self) -> bool
     return false;
 }
 
+/// Clear all references for all keys in the `Map`.
```
Comment (aep) [inline modules/map/src/lib.zz:410]: needs to be clear that it frees all keys and all values,
at which point any borrows to them are no longer valid

**Review by aep — APPROVED** [review]
nice, just nitpicks from here

File: modules/map/src/main.zz [hunk 6]
```diff
@@ -38,12 +38,24 @@ pub fn main() -> int {
...
+
+    err::assert(amap.count() == 3);
 
     let gimmeback = (A*)amap.get("bob");
     err::assert_safe(gimmeback);
     static_attest(buffer::integrity(&gimmeback->b));
     log::info("%s", gimmeback->b.cstr());
 
+    let free_bytes = amap.p.free_bytes();
+    let free_bytes_after_clear = free_bytes + (amap.count() * sizeof(A));
```
Comment (aep) [inline modules/map/src/main.zz:52]: aye, thats exactly the bug i wanted to catch with the test.
if it equals (amap.count() * sizeof(A)), it didn't free the internal Node structure.
so something is wrong with i_dealloc.

Comment (aep) [reply modules/map/src/main.zz:52]: yup, i_dealloc doesnt actually free the node.

would you like to fix this or merge the PR and leave it to me?

Comment (jwerle) [reply modules/map/src/main.zz:52]: i'll take a look!

File: modules/map/src/main.zz [hunk 7]
```diff
@@ -38,12 +39,31 @@ pub fn main() -> int {
...
     err::assert_safe(gimmeback);
     static_attest(buffer::integrity(&gimmeback->b));
     log::info("%s", gimmeback->b.cstr());
 
+    let free_bytes = amap.p.free_bytes();
+    let mut free_bytes_after_clear = free_bytes;
+
+    for let mut keys = amap.keys(); keys.next(); {
+        let mut keysize = pool::align(keys.key.size);
+        let mut nodesize = pool::align(unsafe<usize>(sizeof(map::Node)));
```
Comment (aep) [inline modules/map/src/main.zz:57]: smart idea to use align here to make sure it matches the overhead of pool

Comment (jwerle) [reply modules/map/src/main.zz:57]: seemed like the right thing to do!

**Comment by aep** [comment]
lgtm thanks

**Pull request merged**